server/core/src/main/java/dev/slimevr/osc/VRCOSCHandler.kt

```diff
@@ -110,122 +111,217 @@ class VRCOSCHandler(
-			try {
-				oscQuerySender = OSCPortOut(InetSocketAddress(addr, oscPortOut))
-				oscQuerySender?.connect()
+		if (!config.enabled) {
+			closeOscQuerySender()
+			return
+		}
+
+		try {
+			// If we already have the best matching client
+			if (oscQuerySender != null && oscQueryIpMatch) {
+				return
+			}
+
+			val addr = InetAddress.getByName(oscIP)
+			val ipMatch = ipEquals(addr, this.oscIp)
+
+			// If we already have an OSC sender
+			if (oscQuerySender != null) {
+				val portMatch = oscPortOut == this.oscPortOut
+				// Original IP will be matching because of the check done earlier
+				val origPortMatch = oscQueryPortOut == this.oscPortOut
+
+				// If the IP doesn't match and (the port doesn't match or the original
+				//  port did match), ignore this client
+				if (!ipMatch && (!portMatch || origPortMatch)) {
+					return
+				}
+
+				// If this is the same address as what we're already sending to, keep
+				//  using the same OSC sender
+				if (ipEquals(addr, oscQueryIp)) {
+					return
+				}
+
+				// So if the IP matches or the port matches and the original one didn't,
+				//  we will select this new address for OSC
+			}
+
+			closeOscQuerySender()
+
+			if (ipMatch) {
+				LogManager.info("[VRCOSCHandler] OSCQuery sender sending to port $oscPortOut at address $oscIP (matches configured address)")
+			} else {
 				LogManager.info("[VRCOSCHandler] OSCQuery sender sending to port $oscPortOut at address $oscIP")
-			} catch (e: IOException) {
-				LogManager.severe("[VRCOSCHandler] Error connecting to port $oscPortOut at the address $oscIP: $e")
 			}
+			oscQuerySender = OSCPortOut(InetSocketAddress(addr, oscPortOut))
+			oscQueryIp = addr
+			oscQueryIpMatch = ipMatch
+			oscQueryPortOut = oscPortOut
+
+			// Avoids additional security checks, but not necessary for UDP, therefore
+			//  isConnected doesn't really indicate that we are actively "connected"
+			oscQuerySender?.connect()
+		} catch (e: IOException) {
+			LogManager.severe("[VRCOSCHandler] Error connecting OSCQuery sender to port $oscPortOut at the address $oscIP: $e")
+			closeOscQuerySender()
 		}
 	}
 
 	/**
-	 * Close/remove the osc query sender
+	 * Close/remove the OSC sender
 	 */
-	fun closeOscQuerySender(newState: Boolean) {
-		oscQuerySender?.let {
-			try {
-				it.close()
-				oscQuerySender = null
-				oscQuerySenderState = newState
-			} catch (e: IOException) {
-				LogManager.severe("[VRCOSCHandler] Error closing the OSC sender: $e")
-			}
+	fun closeOscSender() {
+		try {
+			oscSender?.close()
+			oscSender = null
+		} catch (e: IOException) {
+			LogManager.severe("[VRCOSCHandler] Error closing the OSC sender: $e")
 		}
 	}
 
-	override fun updateOscReceiver(portIn: Int, args: Array<String>) {
-		// Stop listening
-		val wasListening = oscReceiver != null && oscReceiver!!.isListening
-		if (wasListening) {
-			oscReceiver!!.stopListening()
+	/**
+	 * Close/remove the OSCQuery sender
+	 */
+	fun closeOscQuerySender() {
```

Comment:
if your addOscQuerySender is synchronized, that should be too, probably.

Reply:
Changed to just use VRServer thread, all the rest are on it anyways so we should probably just keep that up